File: Backtracking/78_Subsets/solution78.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> subset;
        
        dfs(0, nums, subset, res);
        
        return res;
    }
    
private:
    void dfs(int i, vector<int>& nums, vector<int>& subset, vector<vector<int>>& res) {
        if (i >= nums.size()) {
            res.push_back(subset);
            return;
        }
        
        subset.push_back(nums[i]);
        dfs(i + 1, nums, subset, res);
        
        subset.pop_back();
        dfs(i + 1, nums, subset, res);
    }
};

// Runtime1: 0ms, beats 100%
// Memory1: 8.46MB, beats 91.20%

// Runtime2: 0ms, beats 100%
// Memory2: 8.43MB, beats 91.20%

// Runtime3: 0ms, beats 100%
// Memory3: 8.52MB, beats 76.45%

File: Backtracking/78_Subsets/solution78.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        subset = []

        def dfs(i):
            if i >= len(nums):
                res.append(subset.copy())
                return
            subset.append(nums[i])
            dfs(i + 1)
            subset.pop()
            dfs(i + 1)

        dfs(0)
        return res

# Runtime1: 32ms, beats 84.44%
# Memory1: 16.84MB, beats 25.77%

# Runtime2: 33ms, beats 80.25%
# Memory2: 16.83MB, beats 26.83%

# Runtime3: 33ms, beats 80.25%
# Memory3: 16.74MB, beats 26.83%